stginga/utils.py

```diff
@@ -388,3 +394,171 @@ def scale_image(infile, outfile, zoom_factor, ext=('SCI', 1), clobber=False,
+    taking data quality (DQ) flags into account, and adjust WCS accordingly.
+
+    Both PC and CD matrices are supported. Distortion is not
+    taken into account; therefore, this does not work on an
+    image with ``CTYPE`` that ends in ``-SIP``.
+
+    Output image is a single-extension FITS file with only
+    the given extension header and data.
+
+    .. note::
+
+        WCS transformation provided by Mihai Cara.
+
+        Some warnings are suppressed.
+
+    Parameters
+    ----------
+    infile, outfile : str
+        Input and output filenames.
+
+    zoom_factor : float or sequence
+        See :func:`scipy.ndimage.interpolation.zoom`.
+
+    dq_parser : `DQParser`
+        DQ parser for interpreting DQ flag.
+
+    kernel_width : int
+        See :class:`astropy.convolution.Box2DKernel`.
+        This kernel is applied to *rescaled* image.
+
+    sci_ext, dq_ext : int, str, or tuple
+        Science and DQ extensions to extract, respectively.
+
+    bad_flag : int
+        DQ flag value to indicate bad pixels to exclude from calculations.
+        Compound flag is currently not supported.
+
+    ignore_edge_pixels : int
+        Ignore these number of pixels along the edges before rescaling.
+        The default value of 4 is for the reference pixels on JWST NIRCam
+        detectors.
+
+    overwrite : bool
+        If `True`, overwrite existing output file.
+
+    debug : bool
+        If `True`, print extra information to screen.
+
+    Raises
+    ------
+    ValueError
+        Invalid data or WCS.
+
+    """
+    from photutils.utils import ShepardIDWInterpolator
+
+    if not overwrite and os.path.exists(outfile):  # pragma: no cover
+        if debug:
+            warnings.warn('{0} already exists'.format(outfile),
+                          AstropyUserWarning)
+        return  # Instead of raising error at the very end
+
+    with fits.open(infile) as pf:
+        prihdr = pf['PRIMARY'].header
+        hdr = pf[sci_ext].header
+        data = pf[sci_ext].data
+        dq = pf[dq_ext].data
+
+    if data.ndim != 2:  # pragma: no cover
+        raise ValueError('Unsupported ndim={0}'.format(data.ndim))
+
+    # Inherit some keywords from primary header
+    for key in ('ROOTNAME', 'TARGNAME', 'INSTRUME', 'DETECTOR',
+                'FILTER', 'PUPIL', 'DATE-OBS', 'TIME-OBS'):
+        if (key in hdr) or (key not in prihdr):
+            continue
+        hdr[key] = prihdr[key]
+
+    dqs_by_flags = dq_parser.interpret_array(dq)
+
+    # Edge pixels
+    iy_max = data.shape[0] - ignore_edge_pixels
+    ix_max = data.shape[1] - ignore_edge_pixels
+    edge_mask = np.ones_like(dq, dtype=np.bool)
+    edge_mask[ignore_edge_pixels:iy_max, ignore_edge_pixels:ix_max] = False
+
+    badpix_mask = np.zeros_like(dq, dtype=np.bool)
+    badpix_mask[dqs_by_flags[bad_flag]] = True
+
+    goodpix_mask = np.logical_not(badpix_mask)
+    goodpix_mask = zoom(goodpix_mask, zoom_factor)
```

Comment:
Why even resample the `goodpix_mask` when it can be obtained from the "bad" mask?

Reply:
Because both bad and good masks are ignoring the NIRCam reference pixels along the edges. 😬